inspectkv/inspectkv.go

```diff
@@ -0,0 +1,145 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package inspectkv
+
+import (
+	"io"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/column"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/meta"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/terror"
+)
+
+// DDLInfo is for DDL information.
+type DDLInfo struct {
+	SchemaVer   int64
+	ReorgHandle int64
+	Owner       *model.Owner
+	Job         *model.Job
+}
+
+// GetDDLInfo returns DDL information.
+func GetDDLInfo(txn kv.Transaction) (*DDLInfo, error) {
+	var err error
+	info := &DDLInfo{}
+	t := meta.NewMeta(txn)
+
+	info.Owner, err = t.GetDDLOwner()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	info.Job, err = t.GetDDLJob(0)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	info.SchemaVer, err = t.GetSchemaVersion()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if info.Job == nil {
+		return info, nil
+	}
+
+	info.ReorgHandle, err = t.GetDDLReorgHandle(info.Job)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return info, nil
+}
+
+func next(data []interface{}) []interface{} {
+	// Add 0x0 to the end of data.
+	return append(data, []byte{})
+}
+
+// ScanIndexData scans the index handles and values in a limited number.
+// It returns data and the next startVals.
+func ScanIndexData(tableIndexPrefix string, idx *column.IndexedCol, txn kv.Transaction,
+	startVals []interface{}, limit int) ([]int64, [][]interface{}, []interface{}, error) {
+	kvIndex := kv.NewKVIndex(tableIndexPrefix, idx.Name.L, idx.ID, idx.Unique)
+	it, _, err := kvIndex.Seek(txn, startVals)
+	if err != nil {
+		return nil, nil, nil, errors.Trace(err)
+	}
+
+	var (
+		count   int
+		handles []int64
+		vals    [][]interface{}
+		curVals []interface{}
+	)
+	for !table.IsLimit(count, limit) {
+		val, h, err1 := it.Next()
+		if terror.ErrorEqual(err1, io.EOF) {
+			return handles, vals, next(curVals), nil
+		} else if err1 != nil {
+			return nil, nil, nil, errors.Trace(err1)
+		}
+		handles = append(handles, h)
+		vals = append(vals, val)
+		count++
+		curVals = val
+	}
+	nextVals, _, err := it.Next()
+	if terror.ErrorEqual(err, io.EOF) {
+		return handles, vals, next(curVals), nil
+	} else if err != nil {
+		return nil, nil, nil, errors.Trace(err)
+	}
+
+	return handles, vals, nextVals, nil
+}
+
+// ScanTableData scans table row handles and column values in a limited number.
+// It returns data and the next startKey.
+func ScanTableData(t table.Table, retriever kv.Retriever, startKey string, limit int) (
+	[]int64, [][]interface{}, string, error) {
+	var handles []int64
+	var data [][]interface{}
+
+	nextKey, err := t.ScanRecords(retriever, startKey, limit, t.Cols(),
```

Comment:
but the handle only promises not to repeat. So it isn't good to get handle + 1 for next key?